TrafficSimulator/Assets/Car/AutoDrive.cs

```diff
@@ -165,33 +193,104 @@ void Q_UpdateTarget()
                     _status = Status.Driving;
                     
                     // Set the target to the one after the target we missed
-                    _targetIndex = (_repositioningTargetIndex + _repositioningOffset + 1) % _lane.Count;
+                    _targetIndex = (_repositioningTargetIndex + _repositioningOffset + 1 + _lane.Count) % _lane.Count;
                     _target = _lane[_targetIndex];
                     
                     // Reset the max speed to the original, and set the acceleration to the max again
                     _vehicleController.maxSpeedForward = _originalMaxSpeed;
                     _vehicleController.throttleInput = 1f;
+                    _vehicleController.brakeInput = 0f;
                 }
             }
             // If the vehicle is driving and the target is in front of us and we are close enough
             else if (_status == Status.Driving && dot > 0 && direction.magnitude <= _targetLookaheadDistance)
             {
                 // Set the target to the next point in the lane
-                _targetIndex = (_targetIndex + 1) % _lane.Count;
+                int nextTargetIndex = (_targetIndex + 1 + _lane.Count) % _lane.Count;
+                if (_targetIndex >= _lane.Count -1)
+                {
+                    _targetIndex = _roadEndBehaviour == RoadEndBehvaiour.Stop ? _targetIndex : nextTargetIndex;
+                }
+                else
+                {
+                    _targetIndex = nextTargetIndex;
+                }
                 _target = _lane[_targetIndex];
             }
+
+            // If the vehicle is closer to the target than the brake distance, brake
+            if (Vector3.Distance(transform.position, Q_GetBrakeTarget()) <= _brakeDistance)
+            {
+                _vehicleController.brakeInput = 1f;
+                _vehicleController.throttleInput = 0f;
+            }
+            // If the vehicle is further away from the target than the brake distance, accelerate
+            else if (Vector3.Distance(transform.position, Q_GetBrakeTarget()) > _brakeDistance)
+            {
+                _vehicleController.brakeInput = 0f;
+                _vehicleController.throttleInput = 1f;
+            }
+        }
+
+        void Q_UpdateBrakeTarget()
+        {
+            // If the brake target is a stop target, do nothing
+            /*
```

Comment:
Remove commented code.
If you need it later you can find it in the commit history